File: petties_mobile/lib/data/models/booking.dart
/// Booking models for mobile app
/// Matches the backend BookingResponse structure

class BookingResponse {
  final String? bookingId;
  final String? bookingCode;
  final String? clinicId;
  final String? clinicName;
  final String? clinicPhone;
  final String? petId;
  final String? petName;
  final String? petPhotoUrl;
  final String? petSpecies;
  final String? petBreed;
  final double? petWeight;
  final String? ownerId;
  final String? ownerName;
  final String? ownerPhone;
  final String? bookingDate;
  final String? bookingTime;
  final String? status;
  final String? type;
  final String? homeAddress;
  final double? totalPrice;
  final String? notes;
  final String? emrId;
  final List<BookingServiceItem> services;
  final List<BookingPet> pets;
  // Staff info
  final String? assignedStaffName;
  final String? assignedStaffAvatarUrl;

  BookingResponse({
    this.bookingId,
    this.bookingCode,
    this.clinicId,
    this.clinicName,
    this.clinicPhone,
    this.petId,
    this.petName,
    this.petPhotoUrl,
    this.petSpecies,
    this.petBreed,
    this.petWeight,
    this.ownerId,
    this.ownerName,
    this.ownerPhone,
    this.bookingDate,
    this.bookingTime,
    this.status,
    this.type,
    this.homeAddress,
    this.totalPrice,
    this.notes,
    this.emrId,
    this.services = const [],
    this.pets = const [],
    this.assignedStaffName,
    this.assignedStaffAvatarUrl,
  });

  factory BookingResponse.fromJson(Map<String, dynamic> json) {
    // Parse pets list
    final petsList = (json['pets'] as List<dynamic>?)
            ?.map((e) => BookingPet.fromJson(e))
            .toList() ??
        [];

    // Flatten services from all pets if services is not directly provided or empty
    // But since the new API provides services inside pets, we should flatten them to maintain compatibility
    // with existing UI that uses `services` field.
    List<BookingServiceItem> allServices = [];
    if (json['services'] != null) {
      allServices = (json['services'] as List<dynamic>)
          .map((e) => BookingServiceItem.fromJson(e))
          .toList();
    } else {
      for (final pet in petsList) {
        allServices.addAll(pet.services);
      }
    }

    return BookingResponse(
      bookingId: json['bookingId'],
      bookingCode: json['bookingCode'],
      clinicId: json['clinicId'],
      clinicName: json['clinicName'],
      clinicPhone: json['clinicPhone'],
      petId: json['petId'],
      petName: json['petName'],
      petPhotoUrl: json['petPhotoUrl'],
      petSpecies: json['petSpecies'],
      petBreed: json['petBreed'],
      petWeight: (json['petWeight'] as num?)?.toDouble(),
      ownerId: json['ownerId'],
      ownerName: json['ownerName'],
      ownerPhone: json['ownerPhone'],
      bookingDate: json['bookingDate'],
      bookingTime: json['bookingTime'],
      status: json['status'],
      type: json['type'],
      homeAddress: json['homeAddress'],
      totalPrice: (json['totalPrice'] as num?)?.toDouble(),
      notes: json['notes'],
      emrId: json['emrId'],
      services: allServices,
      pets: petsList,
      assignedStaffName: json['assignedStaffName'],
      assignedStaffAvatarUrl: json['assignedStaffAvatarUrl'],
    );
  }
}

class BookingPet {
  final String? petId;
  final String? petName;
  final List<BookingServiceItem> services;

  BookingPet({
    this.petId,
    this.petName,
    this.services = const [],
  });

  factory BookingPet.fromJson(Map<String, dynamic> json) {
    return BookingPet(
      petId: json['petId'],
      petName: json['petName'],
      services: (json['services'] as List<dynamic>?)
              ?.map((e) => BookingServiceItem.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class BookingServiceItem {
  final String? bookingServiceId;
  final String? serviceId;
  final String? serviceName;
  final String? serviceCategory;
  final double? price;
  final double? basePrice;
  final double? weightPrice;
  final int? slotsRequired;
  final String? assignedStaffId;
  final String? assignedStaffName;
  final String? assignedStaffAvatarUrl;
  final String? assignedStaffSpecialty;
  final String? petId;
  final String? petName;
  final String? scheduledStartTime;
  final String? scheduledEndTime;
  final int? durationMinutes;
  final bool? isAddOn;

  BookingServiceItem({
    this.bookingServiceId,
    this.serviceId,
    this.serviceName,
    this.serviceCategory,
    this.price,
    this.basePrice,
    this.weightPrice,
    this.slotsRequired,
    this.assignedStaffId,
    this.assignedStaffName,
    this.assignedStaffAvatarUrl,
    this.assignedStaffSpecialty,
    this.petId,
    this.petName,
    this.scheduledStartTime,
    this.scheduledEndTime,
    this.durationMinutes,
    this.isAddOn,
  });

  factory BookingServiceItem.fromJson(Map<String, dynamic> json) {
    return BookingServiceItem(
      bookingServiceId: json['bookingServiceId'],
      serviceId: json['serviceId'],
      serviceName: json['serviceName'],
      serviceCategory: json['serviceCategory'],
      price: (json['price'] as num?)?.toDouble(),
      basePrice: (json['basePrice'] as num?)?.toDouble(),
      weightPrice: (json['weightPrice'] as num?)?.toDouble(),
      slotsRequired: json['slotsRequired'],
      assignedStaffId: json['assignedStaffId'],
      assignedStaffName: json['assignedStaffName'],
      assignedStaffAvatarUrl: json['assignedStaffAvatarUrl'],
      assignedStaffSpecialty: json['assignedStaffSpecialty'],
      petId: json['petId'],
      petName: json['petName'],
      scheduledStartTime: json['scheduledStartTime'],
      scheduledEndTime: json['scheduledEndTime'],
      durationMinutes: json['durationMinutes'],
      isAddOn: json['isAddOn'],
    );
  }
}

/// ============================================================
/// STAFF HOME SUMMARY MODELS
/// Matches backend StaffHomeSummaryResponse
/// ============================================================

class StaffHomeSummaryResponse {
  final int todayBookingsCount;
  final int pendingCount;
  final int inProgressCount;
  final List<UpcomingBookingDTO> upcomingBookings;

  StaffHomeSummaryResponse({
    required this.todayBookingsCount,
    required this.pendingCount,
    required this.inProgressCount,
    required this.upcomingBookings,
  });

  factory StaffHomeSummaryResponse.fromJson(Map<String, dynamic> json) {
    return StaffHomeSummaryResponse(
      todayBookingsCount: json['todayBookingsCount'] ?? 0,
      pendingCount: json['pendingCount'] ?? 0,
      inProgressCount: json['inProgressCount'] ?? 0,
      upcomingBookings: (json['upcomingBookings'] as List<dynamic>?)
              ?.map((e) => UpcomingBookingDTO.fromJson(e))
              .toList() ??
          [],
    );
  }
}

class UpcomingBookingDTO {
  final String? bookingId;
  final String? bookingCode;
  final String? petName;
  final String? petSpecies;
  final String? petPhotoUrl;
  final String? ownerName;
  final String? ownerPhone;
  final String? bookingDate;
  final String? bookingTime;
  final String? endTime;
  final String? type;
  final String? status;
  final double? totalPrice;
  final String? primaryServiceName;
  final int servicesCount;
  final String? homeAddress;

  UpcomingBookingDTO({
    this.bookingId,
    this.bookingCode,
    this.petName,
    this.petSpecies,
    this.petPhotoUrl,
    this.ownerName,
    this.ownerPhone,
    this.bookingDate,
    this.bookingTime,
    this.endTime,
    this.type,
    this.status,
    this.totalPrice,
    this.primaryServiceName,
    this.servicesCount = 0,
    this.homeAddress,
  });

  factory UpcomingBookingDTO.fromJson(Map<String, dynamic> json) {
    return UpcomingBookingDTO(
      bookingId: json['bookingId'],
      bookingCode: json['bookingCode'],
      petName: json['petName'],
      petSpecies: json['petSpecies'],
      petPhotoUrl: json['petPhotoUrl'],
      ownerName: json['ownerName'],
      ownerPhone: json['ownerPhone'],
      bookingDate: json['bookingDate'],
      bookingTime: json['bookingTime'],
      endTime: json['endTime'],
      type: json['type'],
      status: json['status'],
      totalPrice: (json['totalPrice'] as num?)?.toDouble(),
      primaryServiceName: json['primaryServiceName'],
      servicesCount: json['servicesCount'] ?? 0,
      homeAddress: json['homeAddress'],
    );
  }
}

File: petties-web/src/types/booking.ts
/**
 * Booking Types - Frontend type definitions
 */

// Booking Status enum
export type BookingStatus = 'PENDING' | 'SEARCHING' | 'PENDING_CLINIC_CONFIRM' | 'CONFIRMED' | 'IN_PROGRESS' | 'COMPLETED' | 'CANCELLED' | 'NO_SHOW';

// Booking Type enum
export type BookingType = 'IN_CLINIC' | 'HOME_VISIT' | 'SOS';

// Service item in booking
export interface BookingServiceItem {
    bookingServiceId?: string; // BookingServiceItem ID (junction table PK) - for reassignment
    serviceId: string;
    serviceName: string;
    serviceCategory: string;
    price: number;
    slotsRequired: number;
    durationMinutes: number;

    // Pricing breakdown fields
    basePrice?: number;      // Original service base price
    weightPrice?: number;    // Price tier based on pet weight

    // Assigned staff for this specific service
    assignedStaffId?: string;
    assignedStaffName?: string;
    assignedStaffAvatarUrl?: string;
    assignedStaffSpecialty?: string;

    // Scheduled time for this service
    scheduledStartTime?: string; // For display
    scheduledEndTime?: string;   // For display
    isAddOn?: boolean;           // Arising service

    // Pet this service is for (multi-pet booking)
    petId?: string;
    petName?: string;
}

// Main Booking interface
export interface Booking {
    bookingId: string;
    bookingCode: string;
    emrId?: string; // Linked EMR ID if exists

    // Pet info
    petId: string;
    petName: string;
    petSpecies: string;
    petBreed: string;
    petAge: string;
    petPhotoUrl?: string;
    petWeight?: number; // Weight in kg

    // Owner info
    ownerId: string;
    ownerName: string;
    ownerPhone: string;
    ownerEmail: string;
    ownerAvatarUrl?: string;
    ownerAddress?: string; // Pet owner's registered address

    // Clinic info
    clinicId: string;
    clinicName: string;
    clinicAddress?: string;

    // Staff info
    assignedStaffId?: string;
    assignedStaffName?: string;
    assignedStaffSpecialty?: string;
    assignedStaffAvatarUrl?: string;

    // Payment info
    paymentStatus?: string; // PENDING, PAID, REFUNDED, FAILED
    paymentMethod?: string; // CASH, QR, CARD
    paymentDescription?: string; // SePay payment description for QR matching
    qrImageUrl?: string; // QR code image URL (only for QR + PENDING)

    // Booking info
    bookingDate: string;
    bookingTime: string;
    type: BookingType;
    status: BookingStatus;
    totalPrice: number;
    notes?: string;
    services: BookingServiceItem[];

    /**
     * List of pets in booking (multi-pet), each with its own services.
     */
    pets?: PetInBookingSummary[];

    // Home visit info
    homeAddress?: string;
    homeLat?: number;
    homeLong?: number;
    distanceKm?: number;
    distanceFee?: number; // Home visit fee (pricePerKm × distanceKm) applied once
    sosFee?: number;      // SOS emergency fee

    // Timestamps
    createdAt: string;
}

/**
 * Summary of a pet in a booking (multi-pet) with its services.
 */
export interface PetInBookingSummary {
    petId: string;
    petName: string;
    services: BookingServiceItem[];
}

// Create booking request
export interface CreateBookingRequest {
    petId: string;
    clinicId: string;
    bookingDate: string;
    bookingTime: string;
    type: BookingType;
    serviceIds: string[];
    homeAddress?: string;
    homeLat?: number;
    homeLong?: number;
    notes?: string;
}

// Confirm booking request
export interface ConfirmBookingRequest {
    assignedStaffId?: string;
    selectedStaffId?: string; // For manual staff selection dropdown
    managerNotes?: string;
}

// Status badge colors - Petties Amber-Stone Theme
export const BOOKING_STATUS_CONFIG: Record<BookingStatus, { label: string; bgColor: string; textColor: string }> = {
    PENDING: { label: 'Chờ xác nhận', bgColor: '#F5F5F4', textColor: '#44403C' }, // stone-100 / stone-700
    SEARCHING: { label: 'Đang tìm kiếm', bgColor: '#FEF3C7', textColor: '#D97706' }, // amber-100 / amber-600
    PENDING_CLINIC_CONFIRM: { label: 'Chờ phòng khám', bgColor: '#FEF3C7', textColor: '#D97706' }, // amber-100 / amber-600
    CONFIRMED: { label: 'Đã xác nhận', bgColor: '#FFFBEB', textColor: '#B45309' }, // amber-50 / amber-700
    IN_PROGRESS: { label: 'Đang thực hiện', bgColor: '#FEF3C7', textColor: '#D97706' }, // amber-100 / amber-600 (primary)
    COMPLETED: { label: 'Hoàn thành', bgColor: '#DCFCE7', textColor: '#16A34A' }, // green-100 / green-600
    CANCELLED: { label: 'Đã hủy', bgColor: '#F5F5F4', textColor: '#57534E' }, // stone-100 / stone-600
    NO_SHOW: { label: 'Không đến', bgColor: '#F5F5F4', textColor: '#57534E' }, // stone-100 / stone-600
};

// Booking type config with colors
export const BOOKING_TYPE_CONFIG: Record<BookingType, { label: string; bgColor: string; textColor: string }> = {
    IN_CLINIC: { label: 'Tại phòng khám', bgColor: '#86EFAC', textColor: '#166534' },
    HOME_VISIT: { label: 'Khám tại nhà', bgColor: '#FDE68A', textColor: '#92400E' },
    SOS: { label: 'SOS cấp cứu', bgColor: '#FCA5A5', textColor: '#991B1B' },
};

// Booking type labels (backward compatibility)
export const BOOKING_TYPE_LABELS: Record<BookingType, string> = {
    IN_CLINIC: 'Tại phòng khám',
    HOME_VISIT: 'Khám tại nhà',
    SOS: 'SOS cấp cứu',
};

// Service category labels (Vietnamese)
export const SERVICE_CATEGORY_LABELS: Record<string, string> = {
    CHECK_UP: 'Khám tổng quát',
    VACCINATION: 'Tiêm phòng',
    SURGERY: 'Phẫu thuật',
    DENTAL: 'Nha khoa',
    DERMATOLOGY: 'Da liễu',
    INTERNAL_MEDICINE: 'Nội khoa',
    EMERGENCY: 'Cấp cứu',
    GROOMING_SPA: 'Làm đẹp & Spa',
    BOARDING: 'Trông giữ',
    OTHER: 'Khác',
};

// Payment status labels
export const PAYMENT_STATUS_LABELS: Record<string, { label: string; color: string }> = {
    PENDING: { label: 'Chờ thanh toán', color: '#FCD34D' },
    PAID: { label: 'Đã thanh toán', color: '#4ADE80' },
    REFUNDED: { label: 'Đã hoàn tiền', color: '#93C5FD' },
    FAILED: { label: 'Thất bại', color: '#FCA5A5' },
};

// Payment method labels
export const PAYMENT_METHOD_LABELS: Record<string, string> = {
    CASH: 'Tiền mặt',
    QR: 'QR Code',
    CARD: 'Thẻ tín dụng',
};

// Staff Specialty labels (Vietnamese)
export const STAFF_SPECIALTY_LABELS: Record<string, string> = {
    VET: 'Bác sĩ thú y',
    GROOMER: 'Nhân viên Grooming',
    // Legacy (sau migration DB chỉ còn VET | GROOMER)
    VET_GENERAL: 'Bác sĩ thú y',
    VET_SURGERY: 'Bác sĩ thú y',
    VET_DENTAL: 'Bác sĩ thú y',
    VET_DERMATOLOGY: 'Bác sĩ thú y',
};

// Available staff response for reassignment
export interface AvailableStaffResponse {
    staffId: string;
    staffName: string;
    avatarUrl?: string;
    specialty?: string;
    available: boolean;
    bookedCount: number;
    availableSlots: string[]; // List of available slot times (HH:mm)
    unavailableReason?: string; // If not available, reason why
}

// Reassign staff request
export interface ReassignStaffRequest {
    newStaffId: string;
}

// ========== STAFF AVAILABILITY CHECK ==========

// Service availability in a booking
export interface ServiceAvailability {
    bookingServiceId: string;
    serviceName: string;
    serviceCategory: string;
    requiredSpecialty: string;
    requiredSpecialtyLabel: string;
    price: number;
    hasAvailableStaff: boolean;
    suggestedStaffId?: string;
    suggestedStaffName?: string;
    suggestedStaffAvatarUrl?: string;
    suggestedStaffSpecialty?: string;
    unavailableReason?: string;
}

// Alternative time slot suggestion
export interface AlternativeTimeSlot {
    specialty: string;
    specialtyLabel: string;
    date: string; // ISO date format
    availableTimes: string[]; // HH:mm format
    staffName: string;
    staffId: string;
}

// Staff availability check response
export interface StaffAvailabilityCheckResponse {
    allServicesHaveStaff: boolean;
    services: ServiceAvailability[];
    alternativeTimeSlots: AlternativeTimeSlot[];
    priceReductionIfRemoved: number;
}

// Confirm booking request with options
export interface ConfirmBookingWithOptionsRequest {
    assignedStaffId?: string;
    managerNotes?: string;
    allowPartial?: boolean;
    removeUnavailableServices?: boolean;
}
